src/definitions.rs

```diff
@@ -92,114 +102,391 @@ impl From<FieldPath> for RichTerm {
+            .map(|elt| match elt {
+                SchemaPointerElt::Properties(name) => Some(name.clone()),
+                _ => None,
+            })
+            .collect();
+
+        FieldPath::try_from(stripped?).ok()
+    }
+
+    /// Returns a Nickel term that accesses the value pointed by `self` by looking it up either in
+    /// the references environment or directly in the final contract for pure property paths.
+    pub fn access(&self, usage: RefUsageContext) -> RichTerm {
+        match (self.try_as_field_path(), usage) {
+            // The case of pure property paths is special, as we access them directly from within
+            // the final contract, instead of looking them up in the references environment.
+            (Some(field_path), RefUsageContext::Contract) => field_path.into(),
+            _ => static_access(ENVIRONMENT_ID, [self.nickel_uid(usage).as_str()]),
+        }
+    }
+
+    /// Returns a single mangled string uniquely identifying this pointer and its usage. This name
+    /// is used to store the reference in (and load it from) the references environment.
+    pub fn nickel_uid(&self, usage: RefUsageContext) -> String {
+        let path = self
+            .path
+            .iter()
+            .map(|elt| elt.to_string())
+            .collect::<Vec<_>>()
+            .join("/");
+
+        format!("{MANGLING_PREFIX}:{path}!{usage}")
+    }
+
+    /// Returns `true` if the path is composed only of properties.
+    pub fn is_only_props(&self) -> bool {
+        self.path
+            .iter()
+            .all(|elt| matches!(elt, SchemaPointerElt::Properties(_)))
+    }
+
+    /// Returns the subschema pointed by `self` in the given schema object.
+    ///
+    /// # Return values
+    ///
+    /// - Returns `Some(subschema)` if the path exists in the schema and points to `subschema`.
+    /// - Returns `None` if the path does not exist in the schema or the path is empty.
     ///
-    /// For example, running [Self::try_as_field_path] on a JSON pointer
-    /// `/properties/foo/properties/bar` will return the field path `["foo", "bar"]`.
-    fn try_as_field_path(&self) -> Option<FieldPath> {
-        let mut it = self.path.iter();
-        let mut result = Vec::with_capacity(self.path.len() / 2);
-
-        // We expect that the path can be grouped as a sequence of two elements, where the first
-        // one is always `properties`, and the second one corresponds to the property name.
-        while let Some(part) = it.next() {
-            if part != "properties" {
-                return None;
+    /// Note: it looks like we could return the original value upon empty path, but there's a type
+    /// mismatch: we get a `SchemaObject` reference, and we must return a `Schema` reference. We
+    /// can't convert between the two (we can convert between the owned variants easily, but not
+    /// for references). Since we can special case empty paths before calling to [Self::resolve] if
```

Comment:
```suggestion
    /// for references). Since we can special case empty paths before calling [Self::resolve] if
```

Reply:
I promise one day I'll stop making this error and you'll be free of your duty :stuck_out_tongue: